doc/specs/stdlib_selection.md

```diff
@@ -0,0 +1,336 @@
+to quickly find the median of an array, or some other percentile.
+
+The Fortran Standard Library therefore provides a module, `stdlib_selection`,
+which implements selection algorithms.
+
+## Overview of the module
+
+The module `stdlib_selection` defines two generic subroutines:
+* `select` is used to find the k-th smallest entry of an array. The input
+array is also modified in-place, and on return will be partially sorted
+such that `all(array(1:k) <= array(k)))`  and `all(array(k) <= array((k+1):size(array)))`.
+The user can optionally specify `left` and `right` indices to constrain the search
+for the k-th smallest value. This can be useful if you have previously called `select`
+to find a smaller or larger rank (that will have led to partial sorting of
+`array`, thus implying some constraints on the location).
+
+* `arg_select` is used to find the index of the k-th smallest entry of an array.
+In this case the input array is not modified, but the user must provide an
+input index array with the same size as `array`, having indices that are a permutation of
+`1:size(array)`, which is modified instead. On return the index array is modified
+such that `all(array(index(1:k)) <= array(index(k)))` and `all(array(k) <= array(k+1:size(array)))`.
+The user can optionally specify `left` and `right` indices to constrain the search
+for the k-th smallest value. This can be useful if you have previously called `arg_select`
+to find a smaller or larger rank (that will have led to partial sorting of
+`index`, thus implying some constraints on the location).
+
+### Licensing
+
+The Fortran Standard Library is distributed under the MIT
+License. It is worth noting that  `select` and `arg_select`
+were derived using some code from quickSelect in the Coretran library, by Leon Foks,
+https://github.com/leonfoks/coretran. Leon Foks has given permission for the code here
+to be released under stdlib's MIT license.
+
+## `select` - find the k-th smallest value in an input array
+
+### Status
+
+Experimental
+
+### Description
+
+Returns the k-th smallest value of `array(:)`, and also partially sorts `array(:)`
+such that `all(array(1:k) <= array(k))`  and `all(array(k) <= array((k+1):size(array)))`
+
+### Syntax
+
+`call [[stdlib_selection(module):select(interface)]]( array, k, kth_smallest [, left, right ] )`
+
+### Class
+
+Generic subroutine.
+
+### Arguments
+
+`array` : shall be a rank one array of any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`,
+`real(sp)`, `real(dp)`, `real(qp)`. It is an `intent(inout)` argument. 
+
+`k`: shall be a scalar with any of the types:
+`integer(int8)`, `integer(int16)`, `integer(int32)`, `integer(int64)`. It
+is an `intent(in)` argument. We search for the `k`-th smallest entry of `array(:)`.
+
+`kth_smallest`: shall be a scalar with the same type as `array`. On return it contains
+the k-th smallest entry of `array(:)`.
```

Comment:
The type of `intent` is missing

Reply:
Good catch -- fixed in a commit just now.